ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Counting.hs

```diff
@@ -0,0 +1,378 @@
+pattern Contains x y <- UnsafeContains x y
+
+{-# COMPLETE Contains #-}
+
+forgetWindow :: Contains elem outer inner -> Some.Forgotten (Index outer elem, Index outer elem)
+forgetWindow win = Some.forgotten (windowStart win, windowLast win)
+
+-- | Converts and index of a window into an index in the containing sequence.
+frWin :: Contains elem outer inner -> Index inner elem -> Index outer elem
+frWin (Contains (Count i) _n) (Count j) = Count (i .+ j)
+
+-- | Converts a count of elements in a window to a count of elements in the
+-- containing sequence.
+frWinVar :: Contains elem outer inner -> Var inner x -> Var outer x
+frWinVar _ (Count x) = Count x
+
+toWin :: Contains elem outer inner -> Index outer elem -> Maybe (Index inner elem)
+{-# INLINE toWin #-}
+toWin (Contains (Count i) (Count n)) (Count j) = if i <= j && j < i .+ n then Just (Count (j .- i)) else Nothing
+
+windowSize :: Contains elem outer inner -> Size inner elem
+windowSize (Contains _i (Count n)) = Count n
+
+windowStart :: Contains elem outer inner -> Index outer elem
+windowStart win = frWin win (Count 0)
+
+windowLast :: Contains elem outer inner -> Index outer elem
+windowLast win = frWin win $ lastIndex $ windowSize win
+
+-- | 'Contains' is a 'Data.Semigroupoid.Semigroupoid'
+joinWin :: Contains elem outer mid -> Contains elem mid inner -> Contains elem outer inner
+{-# INLINE joinWin #-}
+joinWin win win2 = UnsafeContains (frWin win $ windowStart win2) (windowSize win2)
+
+data SomeWindow (lbl :: klbl) (outer :: Type) (elem :: kelem) =
+    forall (skolem :: Type).
+    SomeWindow
+        !(Proxy skolem)
+        !(Contains elem outer (Win lbl skolem))
+
+instance Eq (SomeWindow lbl outer elem) where
+    SomeWindow _l1 l2 == SomeWindow _r1 r2 =
+      forgetWindow l2 == forgetWindow r2
+
+instance Show (SomeWindow lbl outer elem) where
+    showsPrec p (SomeWindow prx win) =
+        Some.runShowsPrec p
+      $ Some.showCtor SomeWindow "SomeWindow"
+          `Some.showArg` prx
+          `Some.showArg` win
+
+instance Read (SomeWindow lbl outer elem) where
+    readPrec =
+        Some.runReadPrec
+      $ Some.readCtor SomeWindow "SomeWindow"
+          <*> Some.readArg
+          <*> Some.readArg
+
+-- | @withWindow outerSz lbl offset innerSz@ is a window of length @innerSz@
+-- with name @lbl@ starting at @offset@ in a sequence with length @outerSz@.
+--
+-- If the window doesn't fit in the containing sequence, it is clipped so the
+-- resulting (possibly empty) window is contained.
+--
+-- Note that the window can spill either on the right if @i + innerSz > outerSz@,
+-- or it can spill on the left if @i < 0@, or it can spill on both sides
+-- simultaneously.
+--
+withWindow :: Size outer elem -> Lbl lbl -> Index outer elem -> Size x elem -> SomeWindow lbl outer elem
+withWindow (Count n) _lbl (Count i) (Count m) =
+    SomeWindow Proxy $ UnsafeContains (Count i') (Count m')
+  where
+    i' = min n (max 0 i)
```

Comment:
Hmm, when we have
```haskell
C.SomeWindow Proxy (C.Contains (C.Count i') (C.Count m')) =
  C.withWindow (C.Count n) (C.Lbl :: C.Lbl "test") (C.Count i) (C.Count m)
```
and `n >= 0` and `i >= n`, then we get `i' >= n` which I guess you want to avoid using `min n` as it is beyond the containing window. But we also have `m' == 0`, so the resulting window is empty.

I think there are two options:

 - All empty windows are considered identically (and are in particular vacuously contained in all other windows) regardless of what their index is.
 - Adapt the `prop_withWindow` test to only generate `i < n`, eg naively like this
   ```diff
   diff --git a/ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/ChainGenerator/Tests/Counting.hs b/ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/ChainGenerator/Tests/Counting.hs
   index a32a4eff3..ae329fc6d 100644
   --- a/ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/ChainGenerator/Tests/Counting.hs
   +++ b/ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/ChainGenerator/Tests/Counting.hs
   @@ -19,6 +19,7 @@ tests = [
    
    prop_withWindow :: QC.NonNegative Int -> Int -> Int -> QC.Property
    prop_withWindow (QC.NonNegative n) i m =
   +    (i < n) QC.==>
        case C.withWindow (C.Count n) (C.Lbl :: C.Lbl "test") (C.Count i) (C.Count m) of
            C.SomeWindow Proxy (C.Contains (C.Count i') (C.Count m')) ->
                QC.counterexample (show (i', m'))
   ```

Reply:
I am thinking of 
```diff
--- a/ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/ChainGenerator/Tests/Counting.hs
+++ b/ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/ChainGenerator/Tests/Counting.hs
@@ -25,7 +25,9 @@ prop_withWindow (QC.NonNegative n) i m =
           $ QC.conjoin [
                 if i < 0
                 then QC.counterexample "neg i" $ i' QC.=== 0
-                else QC.counterexample "nonneg i" $ i' QC.=== i
+                else if i > n
+                     then QC.counterexample "nonneg i" $ i' QC.=== n
+                     else QC.counterexample "too large i" $ i' QC.=== i
               , if m < 0
                 then QC.counterexample "neg m" $ m' QC.=== 0
                 else QC.counterexample "nonneg m" $ min (n - 1) i' + m' <= n
```

Reply:
> I am thinking of ...

Adopted this in ccc5a40b for now.